### ISSUE TITLE
Issue with requesting definitions for go components with url encoding in namespaces

### ISSUE THREAD

**Issue by nellshamrell** [clearlydefined/service#884]
Looks like there is a bug in production with go components that did not show up in development - I think because we have more dependencies in the production environment.

Requests for definitions without url encoding in the namespace, such as https://clearlydefined.io/definitions/go/golang/rsc.io/quote/v1.5.2 are working just fine

However, requests for definitions with url encoding such as https://clearlydefined.io/definitions/go/golang/github.com%2fgorilla/mux/v1.7.3 are not showing definition information.

**Comment by nellshamrell** [comment]
Some notes:

These three requests (which are done as part of the definition request in the UI) initially return 404s from the disk cache,  but later succeed with 204s.

* https://api.clearlydefined.io/definitions/go/golang/github.com%2fgorilla/mux/v1.7.3?expand=prs&matchCasing=false
* https://api.clearlydefined.io/curations/go/golang/github.com%2fgorilla/mux/v1.7.3?expand=prs
* https://api.clearlydefined.io/suggestions/go/golang/github.com%2fgorilla/mux/v1.7.3

**Comment by nellshamrell** [comment]
It also appears that https://clearlydefined.io/definitions/go/golang/github.com%2fgorilla/mux/v1.7.3 is not triggering a request to the harvest

**Comment by nellshamrell** [comment]
I believe the prod instance is handling the url encoding/decoding differently from how it was being handled locally. Continuing to troubleshoot.

**Comment by nellshamrell** [comment]
Created a harvest directly with:

```bash
curl -X POST "https://dev-api.clearlydefined.io/harvest" -H "accept: */*" -H "Content-Type: application/json" -d "[{\"tool\":\"package\",\"coordinates\":\"go/golang/github.com%2fgorilla/mux/v1.7.3\"}]"
```

The harvest appears to have run fine (looking at the azure logs, all the tools (ClearlyDefined, Licensee, and Scancode) ran successfully)

**Comment by nellshamrell** [comment]
Yet we still aren't able to see the definition with:

https://clearlydefined.io/definitions/go/golang/github.com%2fgorilla/mux/v1.7.3

or 

```bash
 curl https://dev-api.clearlydefined.io/definitions/go/golang/github.com%2fgorilla/mux/v1.7.3

{"error":{"code":"404","message":"An error has occurred","innererror":{"status":404,"name":"Error","message":"Not Found","stack":"Error: Not Found\n    at requestHandler (/opt/service/app.js:158:17)\n    at Layer.handle [as handle_request] (/opt/service/node_modules/express/lib/router/layer.js:95:5)\n    at trim_prefix (/opt/service/node_modules/express/lib/router/index.js:317:13)\n    at /opt/service/node_modules/express/lib/router/index.js:284:7\n    at Function.process_params (/opt/service/node_modules/express/lib/router/index.js:335:12)\n    at next (/opt/service/node_modules/express/lib/router/index.js:275:10)\n    at /opt/service/node_modules/express/lib/router/index.js:635:15\n    at next (/opt/service/node_modules/express/lib/router/index.js:260:14)\n    at Function.handle (/opt/service/node_modules/express/lib/router/index.js:174:3)\n    at router (/opt/service/node_modules/express/lib/router/index.js:47:12)"}}}
```

**Comment by nellshamrell** [comment]
Did some tracing on:

```bash
curl https://dev-api.clearlydefined.io/definitions/go/golang/github.com%2fgorilla/mux/v1.7.3
```

The request comes through as GET /definitions/go/golang/github.com/gorilla/mux/v1.7.3

Rather than GET /definitions/go/golang/github.com%2fgorilla/mux/v1.7.3

So the url is decoded before it hits the api. Interesting!

**Comment by nellshamrell** [comment]
Working theory - the url may be decoded when it hits the load balancer, before being passed to the api service itself.

**Comment by nellshamrell** [comment]
Interestingly, [this is where things seem to be failing](https://github.com/clearlydefined/service/blob/e2478a1626ebe6cdc13cedefd9afcf48b61b4af9/app.js#L156) when looking at error traces:

**app.js**
```javascript
  // catch 404 and forward to error handler
  const requestHandler = (req, res, next) => {
    const err = new Error('Not Found')
    err.status = 404
    next(err)
  }
  app.use(requestHandler)
```

So it's somehow not finding the definitions route. Digging deeper.

**Comment by nellshamrell** [comment]
The strange thing is that when I attempt to hit this url in a local environment

`http://localhost:3000/definitions/go/golang/github.com/gorilla/mux/v1.7.3` it routes successfully to the definitions route. Interesting.

**Comment by nellshamrell** [comment]
I figured out how to reproduce this locally:

```bash
$ curl http://localhost:4000/definitions/go/golang/github.com/gorilla/test/mux/v1.7.3

{"error":{"code":"404","message":"An error has occurred","innererror":{"status":404,"name":"Error","message":"Not Found","stack":"Error: Not Found\n    at requestHandler (/opt/service/app.js:160:17)\n    at Layer.handle [as handle_request] (/opt/service/node_modules/express/lib/router/layer.js:95:5)\n    at trim_prefix (/opt/service/node_modules/express/lib/router/index.js:317:13)\n    at /opt/service/node_modules/express/lib/router/index.js:284:7\n    at Function.process_params (/opt/service/node_modules/express/lib/router/index.js:335:12)\n    at next (/opt/service/node_modules/express/lib/router/index.js:275:10)\n    at /opt/service/node_modules/express/lib/router/index.js:635:15\n    at next (/opt/service/node_modules/express/lib/router/index.js:260:14)\n    at Function.handle (/opt/service/node_modules/express/lib/router/index.js:174:3)\n    at router (/opt/service/node_modules/express/lib/router/index.js:47:12)"}}}
```

It comes from having too many parameters after definitions. Wondering if I can write some handling for this.

**Comment by nellshamrell** [comment]
This was fixed with #886
